File: examples/basic/asn_lookup.js
/**
 * Basic ASN lookup example
 * Demonstrates how to query RDAP information for an Autonomous System Number
 */

const { RDAPClient } = require('rdapify');

async function main() {
  // Create client
  const client = new RDAPClient({
    cache: true,
    privacy: { redactPII: true },
  });

  try {
    console.log('Querying RDAP information for AS15169 (Google)...\n');

    // Query ASN information (can use number or string with "AS" prefix)
    const result = await client.asn(15169); // or "AS15169"

    console.log('ASN Information:');
    console.log('================');
    console.log(`ASN: AS${result.startAutnum}`);
    if (result.endAutnum && result.endAutnum !== result.startAutnum) {
      console.log(`ASN Range: AS${result.startAutnum} - AS${result.endAutnum}`);
    }
    console.log(`Name: ${result.name || 'N/A'}`);
    console.log(`Country: ${result.country || 'N/A'}`);
    console.log(`Type: ${result.type || 'N/A'}`);
    console.log(`Status: ${result.status?.join(', ') || 'N/A'}`);

    // Display entities (organizations)
    if (result.entities && result.entities.length > 0) {
      console.log('\nOrganizations:');
      result.entities.forEach((entity) => {
        const roles = entity.roles?.join(', ') || 'unknown';
        console.log(`  - ${entity.handle || 'Unknown'} (${roles})`);
      });
    }

    // Display events
    if (result.events && result.events.length > 0) {
      console.log('\nEvents:');
      result.events.slice(0, 3).forEach((event) => {
        const date = new Date(event.date).toLocaleDateString();
        console.log(`  - ${event.type}: ${date}`);
      });
    }

    // Display remarks
    if (result.remarks && result.remarks.length > 0) {
      console.log('\nRemarks:');
      result.remarks.slice(0, 2).forEach((remark) => {
        if (remark.title) {
          console.log(`  - ${remark.title}`);
        }
      });
    }

    // Display metadata
    console.log('\nMetadata:');
    console.log(`  Source: ${result.metadata.source}`);
    console.log(`  Cached: ${result.metadata.cached ? 'Yes' : 'No'}`);
  } catch (error) {
    console.error('\n❌ Error:', error.message);
    if (error.code) {
      console.error('Error Code:', error.code);
    }
    process.exit(1);
  }
}

// Run the example
main().catch(console.error);

File: crates/rdapify/examples/asn_lookup.rs
//! ASN (Autonomous System Number) lookup example.
//!
//! Demonstrates:
//! - Querying ASN information
//! - Handling both "AS15169" and "15169" formats
//! - Processing multiple ASNs
//! - Error handling for invalid ASNs
//!
//! Run with:
//!     cargo run --example asn_lookup

use rdapify::RdapClient;

#[tokio::main]
async fn main() -> rdapify::error::Result<()> {
    println!("=== RDAPify ASN Lookup Example ===\n");

    // Create client with default configuration
    let client = RdapClient::new()?;

    // Query Google's ASN (with "AS" prefix)
    println!("Querying AS15169 (Google)...");
    match client.asn("AS15169").await {
        Ok(response) => {
            println!("  Query: {}", response.query);
            println!("  Name: {:?}", response.name);
            println!("  Country: {:?}", response.country);
            println!("  Start: {:?}", response.start_autnum);
            println!("  End: {:?}", response.end_autnum);
            println!("  Type: {:?}", response.autnum_type);
            println!("  Status: {:?}", response.status);
            println!("  Source: {}", response.meta.source);
        }
        Err(e) => println!("  Error: {}", e),
    }

    println!();

    // Query the same ASN without "AS" prefix
    println!("Querying 15169 (Google, without 'AS' prefix)...");
    match client.asn("15169").await {
        Ok(response) => {
            println!("  Query: {}", response.query);
            println!("  Name: {:?}", response.name);
            println!("  Country: {:?}", response.country);
            println!("  Start: {:?}", response.start_autnum);
            println!("  End: {:?}", response.end_autnum);
            println!("  Type: {:?}", response.autnum_type);
            println!("  Source: {}", response.meta.source);
        }
        Err(e) => println!("  Error: {}", e),
    }

    println!();

    // Query Cloudflare's ASN
    println!("Querying AS13335 (Cloudflare)...");
    match client.asn("AS13335").await {
        Ok(response) => {
            println!("  Query: {}", response.query);
            println!("  Name: {:?}", response.name);
            println!("  Country: {:?}", response.country);
            println!("  Start: {:?}", response.start_autnum);
            println!("  End: {:?}", response.end_autnum);
            println!("  Type: {:?}", response.autnum_type);
            println!("  Status: {:?}", response.status);
            println!("  Source: {}", response.meta.source);
        }
        Err(e) => println!("  Error: {}", e),
    }

    println!();

    // Demonstrate error handling with an invalid ASN
    println!("Querying AS0 (invalid ASN - should error)...");
    match client.asn("AS0").await {
        Ok(response) => {
            println!("  Query: {}", response.query);
            println!("  Name: {:?}", response.name);
        }
        Err(e) => {
            println!("  Error (expected): {}", e);
        }
    }

    println!("\nCache size: {} entries\n", client.cache_size());

    println!("Example completed successfully!");
    Ok(())
}
